### ISSUE TITLE
[platform] Rate limit network usage when doing backup/restore

### ISSUE THREAD

**Issue by bmatican** [yugabyte/yugabyte-db#3851]
I believe currently we do not throttle s3cmd network usage when doing either backup or restore.

Aha! Link: https://yugabyte-test.aha.io/features/PLATFORM-635

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-3851-analysis.md`):

# Analysis: yugabyte/yugabyte-db#3851 (Aha PLATFORM-635)

**Title:** [platform] Rate limit network usage when doing backup/restore
**Filed:** 2020-03-06 · **State:** OPEN · **Labels:** area/platform · **Comments:** 0
**Kind:** control-plane (YBA) feature request → behavior-delta validation (no crash repro applies).
**Validated:** 2026-08-29 on master checkout 46a9f7adfd.

## Verdict
VALIDATED — the 2020 claim ("we do not throttle s3cmd network usage when doing either backup or restore") was accurate when filed, and the requested capability has since been delivered on the supported (YB-Controller) backup path: runtime-settable concurrency throttles (GA, PLAT-5292, 2022-09-03) plus absolute bytes/sec rate limiting (preview since YBA 2025.2.0.0, PLAT-17560, 2025-07-02). Quick verdict: **Fixed** (delivered-with-residual: the B/s cap is disk-I/O-scoped rather than a direct network-bandwidth knob, and the legacy s3cmd script path remains unthrottled).

## Root cause
Missing feature, not a code defect. At filing time the only backup mechanism was the platform-invoked script `managed/devops/bin/yb_backup.py`, whose S3 transfers shell out to s3cmd with no bandwidth flag — `managed/devops/bin/yb_backup.py:636-645` builds the s3cmd argv (`--force --no-check-certificate --config=...` [+ `--disable-multipart`/`--access_token`]) and never passes s3cmd's `--limit-rate`; the sole throughput knob is transfer concurrency (`-j/--parallelism`, default 8, `managed/devops/bin/yb_backup.py:1330`). Backup/restore could therefore saturate node NICs/disks and starve production traffic, exactly as reported.

## Current status (delivery evidence, master 46a9f7adfd)
The platform replaced the script flow with YB-Controller-mediated backups (YBA ≥ 2.16 per `docs/content/stable/yugabyte-platform/back-up-restore-universes/_index.md:37`) and grew a two-generation throttle surface:

1. **Concurrency throttling (GA, since PLAT-5292 `ce6a2f02647e` 2022-09-03):** `max_concurrent_uploads`, `per_upload_num_objects`, `max_concurrent_downloads`, `per_download_num_objects` — `managed/src/main/java/com/yugabyte/yw/forms/YbcThrottleParameters.java:19-33`, gflag names `managed/src/main/java/com/yugabyte/yw/common/gflags/GFlagsUtil.java:192-195`, REST `GET/POST /customers/:cUUID/universes/:uniUUID/ybc_throttle_params` (+`_async`) `managed/src/main/resources/v1.routes:636-638`, presets (default 2, min 1, max cores+1) `managed/src/main/java/com/yugabyte/yw/common/backuprestore/ybc/YbcManager.java:198-224`, UI "Configure Resource Throttling" dialog (`managed/ui/src/components/backupv2/components/BackupThrottleParameters.tsx`).
2. **Absolute bytes/sec rate limiting (preview since YBA 2025.2.0.0, PLAT-17560 `87c19523aff6` 2025-07-02):** `disk_read_bytes_per_sec` (backup upload path) and `disk_write_bytes_per_sec` (restore download path) — `YbcThrottleParameters.java:35-49`, non-restart-settable YBC gflags, min 1 MB/s / 0 = unlimited (`YbcManager.java:209-220`), runtime-conf defaults `ybc.default_disk_io_read/write_bytes_per_sec` (`managed/src/main/java/com/yugabyte/yw/common/config/UniverseConfKeys.java:1852-1867`), YBC-side limiter yugabyte/ybc@09ee27ccd0857a60a6e9a8facb750b00c2cc4eca, UI in MB/s (PLAT-18300 `fe8cdfda25d5`, PLAT-18531 `cb02d1802930`), documented (DOC-1142 `a145264cdbb8`; `docs/content/stable/yugabyte-platform/back-up-restore-universes/back-up-universe-data.md:88-108` — "rate-limit disk throughput during backup upload and restore download to reduce impact on cluster operations").

The B/s cap sits at the disk stage of the same transfer pipeline (backup: disk read → upload; restore: download → disk write), so it bounds network transfer rate as a direct consequence in both directions — the issue's core want (backup/restore cannot starve production) is deliverable today via API/UI at runtime.

**Residual gap:** (a) no *direct* network-bandwidth (egress/ingress MB/s) knob exists — the caps are disk-I/O-scoped by name and mechanism; (b) the literal surface named in the issue (yb_backup.py's s3cmd) never received rate limiting and still has none on master (`yb_backup.py:636-645`), but that path is legacy. No commit ever references GH #3851 (only unrelated Jira DB-3851, `9a63e5bda217`); the issue was never issue-tagged closed.

## Trigger matrix
Behavior-delta matrix (capability probe; predicted from the 2020 issue claim vs observed on master 46a9f7adfd):

| Condition (capability probe) | Predicted (issue era, 2020) | Observed (master 46a9f7adfd) |
|---|---|---|
| s3cmd argv carries a bandwidth flag in yb_backup.py | absent (issue claim) | absent — `yb_backup.py:636-645`, no `--limit-rate` (matches) |
| Script-path throughput knob | none | concurrency only (`-j` default 8, `yb_backup.py:1330`) — no bandwidth cap (matches) |
| Platform API to throttle backup/restore | none | present — `v1.routes:636-638`, since PLAT-5292 2022-09-03 (delta: delivered) |
| Absolute bytes/sec rate limit | none | present (preview 2025.2.0.0) — `disk_read/write_bytes_per_sec`, min 1 MB/s, `YbcThrottleParameters.java:35-49` (delta: delivered) |
| Direct network-bandwidth (MB/s egress) knob | requested | still absent — B/s caps disk-scoped; network bounded only as pipeline consequence (residual) |
| Fix commit referencing GH #3851 | — | none (`gh search commits`: only Jira DB-3851, unrelated) |

All cells observed as predicted from source reading before command execution.

## Reproduction
N/A — control-plane feature request; there is no crash/wrong-result to reproduce. Validation is the in-tree/docs/history behavior delta, reproducible read-only against this repo:
- `gh issue view 3851 --repo yugabyte/yugabyte-db --json title,body,labels,state,createdAt` — claim + open state.
- `rg -n "bandwidth|limit.rate" managed/devops/bin/yb_backup.py` → no matches; read `managed/devops/bin/yb_backup.py:636-645` — s3cmd path still unthrottled.
- Read `managed/src/main/java/com/yugabyte/yw/forms/YbcThrottleParameters.java`, `managed/src/main/resources/v1.routes:636-638`, `managed/src/main/java/com/yugabyte/yw/common/backuprestore/ybc/YbcManager.java:198-224` — delivered throttle surface.
- `gh api "repos/yugabyte/yugabyte-db/commits?path=managed/src/main/java/com/yugabyte/yw/forms/YbcThrottleParameters.java&per_page=100"` and `gh search commits --repo yugabyte/yugabyte-db "disk_read_bytes_per_sec"` / `"3851"` — lineage (local clone is shallow).
- Docs: `docs/content/stable/yugabyte-platform/back-up-restore-universes/back-up-universe-data.md:88-108`.
Full command/output transcript: `/tmp/claude/yb3851-evidence.txt`.

## Versions tested
- Source: master checkout 46a9f7adfd (2026-08-29, shallow); history via GitHub API.
- Feature versions established from evidence: concurrency throttle params since YBA ~2.16 (commit 2022-09-03, PLAT-5292); bytes/sec rate limiting preview since YBA 2025.2.0.0 (PLAT-17560, `@YbaApi(sinceYBAVersion = "2025.2.0.0")`).
- Master affected: the *residual* persists on master (no direct network-MB/s knob; s3cmd path unthrottled); the core capability is present on master.

## Proposed fix direction
None required for the core ask — recommend closing #3851 as delivered by PLAT-5292 + PLAT-17560/PLAT-18300, pointing at the "Configure backup performance parameters" docs. If maintainers deem the residual worth tracking, file a fresh, narrowly-scoped follow-up: (a) an explicit network-bandwidth throttle gflag in YBC (yugabyte/ybc repo — closed-source component; YBA side would add one field to `YbcThrottleParameters.java` + preset + UI, mirroring the PLAT-17560 pattern), and/or (b) `--limit-rate` pass-through for the legacy `yb_backup.py` s3cmd path (one-line argv addition at `yb_backup.py:639` gated by a new CLI arg) if that path is still supported anywhere.

## Test strategy
Platform-side unit tests: `managed/src/test/java/com/yugabyte/yw/common/backuprestore/ybc/YbcManagerTest.java` (throttle preset/get-set coverage) and controller tests for `BackupsController`/`YbcController` exercise the public REST surface (`POST .../ybc_throttle_params`). Any residual-fix would extend `YbcThrottleParameters` round-trip tests plus a `getThrottleParamsMap` preset assertion — all via the public API forms, no test-only production hooks. UI: `managed/ui` jest snapshot for `BackupThrottleParameters.tsx`. YBC-side rate-limiter behavior lives in the closed-source yugabyte/ybc repo and is out of scope here.

## Confidentiality check
Clean. All content derives from the public GitHub issue, public commits, and this open-source repo. No customer identifiers, PII, schemas, secrets, or internal hostnames. The Aha link (PLATFORM-635) appears in the public issue body itself. Sole non-public reference is to the closed-source yugabyte/ybc commit hash, which Yugabyte itself published in the public commit message of `87c19523aff6`.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb3851-evidence.txt`):

yb3851 evidence transcript — yugabyte/yugabyte-db#3851 "[platform] Rate limit network usage when doing backup/restore"
Collected 2026-08-29 on checkout master 46a9f7adfd (shallow clone; history probed via gh api / gh search commits).
All paths relative to /Users/brandonmcnama/src/github.com/yugabyte/yugabyte-db.

================================================================================
[1] Issue snapshot
================================================================================
$ gh issue view 3851 --repo yugabyte/yugabyte-db --json title,body,labels,state,createdAt,comments,closedAt
{"body":"I believe currently we do not throttle s3cmd network usage when doing either backup or restore.\n\nAha! Link: https://yugabyte-test.aha.io/features/PLATFORM-635",
 "closedAt":null,"comments":[],"createdAt":"2020-03-06T02:47:34Z",
 "labels":[{"name":"area/platform","description":"Yugabyte Platform"}],
 "state":"OPEN","title":"[platform] Rate limit network usage when doing backup/restore"}

================================================================================
[2] Legacy script path: s3cmd still has NO bandwidth limit (master 46a9f7adfd)
================================================================================
$ rg -n "s3cmd" managed/devops/bin/yb_backup.py          # 6 hits, none a rate limit
$ rg -n "bandwidth|limit.rate|limitrate" managed/devops/bin/yb_backup.py
(no matches)  -> s3cmd `--limit-rate` is never passed.

managed/devops/bin/yb_backup.py:636-645 (S3BackupStorage._command_list_prefix):
        args = ['s3cmd', '--force', '--no-check-certificate', '--config=%s'
                % self.options.cloud_cfg_file_path]
        if self.options.args.disable_multipart:
            args.append('--disable-multipart')
        if self.options.access_token:
            args.append('--access_token=%s' % self.options.access_token)
-> upload/download argv: put/get/sync + SSE only. No bandwidth flag exists on this path.

Only throughput knob is concurrency:
managed/devops/bin/yb_backup.py:1330  '-j', '--parallelism', type=check_arg_range(1, 100), default=8,
managed/devops/bin/yb_backup.py:1469,2598,3839  ThreadPool(self.args.parallelism)

================================================================================
[3] Modern YBC path: throttle parameters (concurrency, GA since 2022)
================================================================================
managed/src/main/java/com/yugabyte/yw/forms/YbcThrottleParameters.java:19-33
  maxConcurrentUploads   (JSON alias max_concurrent_uploads)   "Max concurrent uploads per node"
  perUploadNumObjects    (per_upload_num_objects)              "Max objects per upload per node"
  maxConcurrentDownloads (max_concurrent_downloads)            "Max concurrent downloads per node"
  perDownloadNumObjects  (per_download_num_objects)            "Max objects per download per node"

Gflag names: managed/src/main/java/com/yugabyte/yw/common/gflags/GFlagsUtil.java:192-197
REST endpoints: managed/src/main/resources/v1.routes:636-638
  GET  /customers/:cUUID/universes/:uniUUID/ybc_throttle_params        BackupsController.getThrottleParams
  POST /customers/:cUUID/universes/:uniUUID/ybc_throttle_params        BackupsController.setThrottleParams
  POST /customers/:cUUID/universes/:uniUUID/ybc_throttle_params_async  YbcController.setThrottleParamsAsync
Async set API: managed/src/main/java/com/yugabyte/yw/controllers/YbcController.java:202-241
Presets: managed/src/main/java/com/yugabyte/yw/common/backuprestore/ybc/YbcManager.java:198-224
  CONCURRENCY_PARAM: default 2, min 1, max numCores+1
  NUM_OBJECTS_PARAM: default numCores/2+1, min 1, max numCores+1
Default max_concurrent_uploads=2: managed/src/main/java/com/yugabyte/yw/forms/ybc/YbcGflags.java:181-183

================================================================================
[4] Modern YBC path: absolute bytes/sec rate limiting (preview since YBA 2025.2.0.0)
================================================================================
managed/src/main/java/com/yugabyte/yw/forms/YbcThrottleParameters.java:35-49
  diskReadBytesPerSecond  (disk_read_bytes_per_sec)  "Disk read bytes per second to throttle disk usage during backups"
      @YbaApi(visibility = PREVIEW, sinceYBAVersion = "2025.2.0.0")
  diskWriteBytesPerSecond (disk_write_bytes_per_sec) "Disk write bytes per second to throttle disk usage during restore"
      @YbaApi(visibility = PREVIEW, sinceYBAVersion = "2025.2.0.0")
Presets: YbcManager.java:209-220  DISK_IO_READ_PARAM / DISK_IO_WRITE_PARAM: min 1024*1024 (1 MB/s), default from runtime conf
Runtime-conf defaults: managed/src/main/java/com/yugabyte/yw/common/config/UniverseConfKeys.java:1852-1867
  ybc.default_disk_io_read_bytes_per_sec  "Default disk IO read bytes per second during backup in Yb-Controller"
  ybc.default_disk_io_write_bytes_per_sec "Default disk IO write bytes per second during restore in Yb-Controller"
UI (MB/s): managed/ui/src/components/backupv2/components/BackupThrottleParameters.tsx:279
  "Min {convertBytesToMB(getPresetValues('disk_read_bytes_per_sec', 'min')).toFixed(0)} MB/s"

================================================================================
[5] Documentation (current stable docs)
================================================================================
docs/content/stable/yugabyte-platform/back-up-restore-universes/back-up-universe-data.md:88-108
  "## Configure backup performance parameters
   You can manage the speed of backup and restore operations and their impact on database
   performance by configuring the following parameters:
   - Parallel uploads per node. ...
   - Buffers per upload per node. ...
   - Disk read/write bytes per second (in MB/s). You can rate-limit disk throughput during
     backup upload and restore download to reduce impact on cluster operations. The minimum
     value is 1 MB/s. To allow backup and restore to use as much throughput as they can
     (that is, no disk-based throttling), enter a value of 0 (the default).
   ...
   1. Click **Advanced** and choose **Configure Throttle Parameters** to display the
      **Configure Resource Throttling** dialog."
docs/content/stable/yugabyte-platform/back-up-restore-universes/_index.md:37
  "Configurable performance parameters and incremental backups are mediated using the
   yb-controller process, which is only available in YugabyteDB Anywhere v2.16 or later..."

================================================================================
[6] Commit lineage (gh api / gh search commits; local clone is shallow)
================================================================================
$ gh api "repos/yugabyte/yugabyte-db/commits?path=managed/src/main/java/com/yugabyte/yw/forms/YbcThrottleParameters.java&per_page=100" (tail)
  87c19523aff6 2025-07-02 [PLAT-17560] YBA changes to accommodate YBC rate limiting
  7fc8d2cb7a1b 2023-04-25 [PLAT-8498] Incorrect throttle parameter values for dedicated nodes/K8s pods
  ba8f8d21757f 2023-04-24 [PLAT-8550] Enable per-file java linter + migrate to latest sbt-java-formatter
  37a930547ff7 2023-04-23 [PLAT-8498] Incorrect throttle param values for K8s pods/dedicated nodes
  ce6a2f02647e 2022-09-03 [PLAT-5292] Allow users to set throttle params in Yb-Controller   <- feature origin

$ gh search commits --repo yugabyte/yugabyte-db "disk_read_bytes_per_sec"
  fe8cdfda25d5 2025-09-01 [PLAT-18300] Add UI option to support disk throttling during backup/restore in YBC throttle params window
  87c19523aff6 2025-07-02 [PLAT-17560] YBA changes to accommodate YBC rate limiting

$ gh search commits --repo yugabyte/yugabyte-db "throttle" (relevant subset)
  cb02d1802930 2025-09-17 [PLAT-18531] Fix UI screen for disk throttle params to use MB (convert and send as bytes to API)
  5d6f90e979e5 2025-10-30 [PLAT-18199][Platform][UI] Make the YBC throttle params upgrade an async API
  55d5100c7fd7 2025-02-05 [PLAT-16356] Add YBC Throttle Params to the universe CRD
  a145264cdbb8 2026-02-06 [DOC-1142] Added disk-based throttling in YBC (#30137)

$ gh api repos/yugabyte/yugabyte-db/commits/87c19523aff6 --jq '.commit.message' (excerpt)
  "This PR adds support to modify rate limiting throttling param on YBC via throttle
   parameters configuration API. The YBC disk rate limiting was introduced as part of this
   commit: https://github.com/yugabyte/ybc/commit/09ee27ccd0857a60a6e9a8facb750b00c2cc4eca
   The following API can now be used to modify the disk related throttle parameters in YBC:
   - disk_read_bytes_per_sec
   - disk_write_bytes_per_sec
   Both of these are non-restart settable gflag on YBC, like all other throttling parameters.
   POST api/v1/customers/{cUUID}/universes/{uniUUID}/ybc_throttle_params"

$ gh search commits --repo yugabyte/yugabyte-db "3851"
  9a63e5bda217 [DB-6875] [DB-3851] YCQLSH: Updating yugabyte-db-thirdparty release version
  -> Jira DB-3851, unrelated. NO commit ever references GitHub issue #3851.

================================================================================
[7] Residual gap (still true on master 46a9f7adfd)
================================================================================
- No direct NETWORK-bandwidth (egress/ingress MB/s) knob exists. The bytes/sec caps are
  disk-I/O-scoped (YbcThrottleParameters.java:35-49); they bound network transfer rate only
  as a consequence of capping the disk stage of the same backup/restore pipeline.
- The literal surface named in the issue (yb_backup.py's s3cmd) never received rate
  limiting and still has none (yb_backup.py:636-645); that path is legacy — throttle
  parameters are mediated by YB-Controller (YBA >= 2.16 per docs).


[PLAT-17560]: https://yugabyte.atlassian.net/browse/PLAT-17560?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
